### DISCUSSION TITLE
Tree widget

### DISCUSSION THREAD

**Discussion by Xeverous** [cycfi/elements#333] [Elements]
This issue is for tracking any decisions before implementing this widget.

Something like this: https://wxpython.org/Phoenix/docs/html/_images/wx.treectrl2.png

**Comment by redtide** [comment]
I'm interested on this but I wonder how it would be implemented; on wxWidgets and Qt there are some associated models.
E.G. I'm interested to use it as a filemanager widget; in Qt you need to associate a QFileSystemModel to a QTreeView.
See also <https://doc.qt.io/qt-5/modelview.html>, this is related also to List|Table Widgets|Views.

**Comment by Xeverous** [comment]
@redtide elements has a very loose coupling of widgets and likely there won't be any models for trees. The best example is progress bar element: it doesn't expect any concrete child elements, it only resizes the foreground element on the constant-size background element. Tree widget would be similar: only adding padding to child elements. I think you can already implement own tree wrapper using existing `vector_composite`.

**Comment by djowel** [comment]
Let's make a gallery factory out of it. It should be easy. I'll see what I can do.

**Comment by djowel** [comment]
> @redtide elements has a very loose coupling of widgets and likely there won't be any models for trees.

Basically, a tree is just a hierarchical vtile (or vgrid if you will) composed of vtiles.

**Comment by djowel** [comment]
> I'm interested on this but I wonder how it would be implemented; on wxWidgets and Qt there are some associated models.

There is no such thing in Elements and I'd want to avoid such abstractions for the sake of straightforward simplicity. What I probably want is to have a Boost.Fusion style adapter for a data structure (call it your model if you will) that can be used to generate a vtile hierarchy from element "view-templates" that are used for specifying how the elements will look like (call it your view if you will). Then finally, a factory function that binds the two and sync the view them when the "model" data structure changes, or vice versa.

I admit it will not be as easy as I initially described it, but I have to remind myself that this is indeed the higher-level scheme that I plan for that will sit above the "gallery" and the basic elements. nevertheless, it is very doable with some internal c++ metaprogramming magic.
